test d3figures

dhvenema opened the pull request.

Commit:
test d3figures

dhvenema commented:
@marcoscaceres Could you have a look at this? It doesn't quite work. Currently, I'm nesting an `<svg>` inside the `<figure>`, perhaps that's not possible? Also, both the width-height attributes and the style attribute assigned to the `<svg>` don't seem to work. 

However, when hoovering the mouse over figure 2, the transform functionaily (as displayed in the element inspector in the browser) does seem to update the x-, y-position and scale, but nothing happens. Any ideas?

Here's an example of the d3 functionality (zoom+pan): https://bl.ocks.org/sgruhier/50990c01fe5b6993e82b8994951e23d0.

marcoscaceres commented on js/geonovum/d3figures.js:
```diff
@@ -0,0 +1,69 @@
+define(
+  ["geonovum/deps/d3",
+   "core/pubsubhub"],
```
note for later: you shouldn't need pubsubhub.

dhvenema replied:
I included it for later, in case I needed to pub some error. I will just remove it.

marcoscaceres commented on js/geonovum/d3figures.js:
```diff
@@ -0,0 +1,69 @@
+   */
+  return {
+    run: function(conf, doc, cb) {
+      var figures = $("figure.scalable");
```
here, you can just do: 
```
var images = doc.querySelectorAll("figure.scalable img"); 
```
Then you don't the stuff on line 15 below.

marcoscaceres replied:
(fixed typo above)

marcoscaceres commented on js/geonovum/d3figures.js:
```diff
@@ -0,0 +1,69 @@
+    run: function(conf, doc, cb) {
+      var figures = $("figure.scalable");
+      
+      for (i = 0; i < figures.length; i++) {
```
note: this will fail at build time, because "i" is not declared using var (javascript has a special strict mode, which ReSpec relies on). Should be:
```
for (var i = 0; i < figures.length; i++) {
```

marcoscaceres commented on js/geonovum/d3figures.js:
```diff
@@ -0,0 +1,69 @@
+        var img = figures[i].getElementsByTagName("img")[0];
+        
+        /* Get and/or set image height */
+        if (img.getAttribute("height")) {
```
note, generally, you want to avoid "getAttribute" and "setAttribute". Just set/get attributes directly. Like:

```js
if (!image.width) {
  image.width = 123;
}
```

marcoscaceres commented on js/geonovum/d3figures.js:
```diff
@@ -0,0 +1,69 @@
+        }
+        
+        /* Create svg element */
+        var node = document.createElement("svg")
```
Ok, you probably don't want to be doing this.

marcoscaceres commented on js/geonovum/d3figures.js:
```diff
@@ -0,0 +1,69 @@
+        if (img.getAttribute("width")) {
+          width = img.getAttribute("width");
+        } else {
+          width = "100%";
```
Value 100% is not a valid value in HTML5.

marcoscaceres commented on js/geonovum/d3figures.js:
```diff
@@ -0,0 +1,69 @@
+          width = "100%";
+          img.setAttribute("width", width);
+        };
+        
```
Note, you want to do this with CSS:

```CSS
figure.scalable img {
   width: 100%; /*probably want auto here instead*/
}
```

marcoscaceres commented:
Just noting, you need to run:

```
npm install --save-dev d3
```

to make sure d3 is added to package.json 

Mode extensive comments coming soon...

dhvenema committed:
Update d3figures.js

dhvenema committed:
Update whitepapertest.html

dhvenema commented:
Updated the d3figures.js. The zoom-pan sort of works. However, currently only the last image is able to zoom-pan and previous images don't work.

marcoscaceres commented on js/geonovum/d3figures.js:
```diff
@@ -0,0 +1,38 @@
+       Array
+        .from(doc.querySelectorAll("figure.scalable img"))
+        .forEach(function(image){
+          svg = d3.select(image.parentElement)
```
Add "var svg ="

marcoscaceres commented on js/geonovum/d3figures.js:
```diff
@@ -0,0 +1,38 @@
+                  .attr("width", ((image.getAttribute("width")) ? image.getAttribute("width") : "100%"))
+                  .attr("height", ((image.getAttribute("height")) ? image.getAttribute("height") : 600))
+                  .style("border", "1px solid black");
+          g = svg.append("g")
```
Add "var g = " here too.

dhvenema replied:
Adding var fixed some of it :)

marcoscaceres reviewed (commented):
Adding "var" should hopefully get it working correctly

marcoscaceres commented:
Can simplify things a bit more, but let's see if it works first

dhvenema committed:
Add files via upload

dhvenema committed:
Update d3figues.js

dhvenema commented:
@marcoscaceres I changed a couple of things and I got the basic zoom+pan functionality working for all `<figure>`. Thanks for the suggestions. 

Now, I also added two buttons: a button to reset the zoom+pan (is commented out),  and a button to enlarge the image, i.e. open it in a new tab (similar to a clickable image that opens in a new tab). Neither of them seem to work properly. Zoom+pan, reset and enlarge is the entire set of functionality I had in mind.

Beside that: the images don't align to the center of the svg-box. 

Any ideas?

marcoscaceres commented on js/geonovum/d3figures.js:
```diff
@@ -0,0 +1,57 @@
+          var reset_button = d3.select(image.parentElement)
+                               .insert("button", ":nth-child(2)")
+                               .attr("class", "")
+                               .attr("type", "button")
```
Nit: in theory, you should not need this line.

marcoscaceres commented:
I'd have to poke around a bit more at the SVG, to see how it works (I'm not familiar at all with SVG, to be honest). 

From here, I would try first testing a very simple SVG file (i.e., just a circle or square), to see if the code is correct. I noticed, for instance, that the deer SVG already contains a "g" element - but I don't know if that is screwing anything up. It might be that the more complex SVG image you are currently using is the problem - and not d3.

Further, it also seems strange that it's necessary to have to recreate the SVG image, and then insert another image into it (which is also an SVG image) to get it to work. Conceptually, I can see why it works, but it doesn't feel like a very clean solution... though it might be ok for now 🤷‍♂️

dhvenema committed:
Update d3figures.js

dhvenema commented:
The zoom+pan works great. The reset doesn't reset completly, it set the zoom and x-y to the initial values, but some internal zoom doesn't seem to get reset. Thus, when moved slightly after the reset, it jumps back to the previous values. 

And, the enlarge button also doesn't work. Even though, it's specified to only activate the new window on click, it immediately opens up (Firefox blocks it). Afterwards, the button doesn't do anything.

@marcoscaceres I have seen some examples of using the SVG structure, but then I would have to make some expections as I also want this to work on PNG or other image formats.

dhvenema committed:
Merge branch 'develop' into d3figures
